source/common/network/tcp_keepalive_option_impl.cc

```diff
@@ -0,0 +1,66 @@
+#include "common/network/tcp_keepalive_option_impl.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+bool TcpKeepaliveOptionImpl::setOption(Network::Socket& socket,
+                                       Network::Socket::SocketState state) const {
+  if (state == Socket::SocketState::PreBind) {
+    return setTcpKeepalive(socket, keepalive_config_.keepalive_probes_,
+                           keepalive_config_.keepalive_time_,
+                           keepalive_config_.keepalive_interval_);
+  }
+  return true;
+}
+
+bool TcpKeepaliveOptionImpl::setTcpKeepalive(Socket& socket, absl::optional<int> keepalive_probes,
+                                             absl::optional<int> keepalive_time,
+                                             absl::optional<int> keepalive_interval) {
+  int error;
+  error = setSocketOption(socket, SOL_SOCKET, ENVOY_SOCKET_SO_KEEPALIVE, 1);
+  if (error != 0) {
+    ENVOY_LOG(warn, "Setting SO_KEEPALIVE on socket failed: {}", strerror(errno));
+    return false;
+  }
+  error = setSocketOption(socket, IPPROTO_TCP, ENVOY_SOCKET_TCP_KEEPCNT, keepalive_probes);
+  if (error != 0) {
+    ENVOY_LOG(warn, "Setting keepalive_probes failed: {}", strerror(errno));
+    return false;
+  }
+  error = setSocketOption(socket, IPPROTO_TCP, ENVOY_SOCKET_TCP_KEEPIDLE, keepalive_time);
+  if (error != 0) {
+    ENVOY_LOG(warn, "Setting keepalive_time failed: {}", strerror(errno));
+    return false;
+  }
+  error = setSocketOption(socket, IPPROTO_TCP, ENVOY_SOCKET_TCP_KEEPINTVL, keepalive_interval);
+  if (error != 0) {
+    ENVOY_LOG(warn, "Setting keepalive_interval failed: {}", strerror(errno));
+    return false;
+  }
+  return true;
+}
+
+int TcpKeepaliveOptionImpl::setSocketOption(Socket& socket, int level,
+                                            Network::SocketOptionName optname,
+                                            absl::optional<int> optional_value) {
+  if (optional_value.has_value()) {
+    return setSocketOption(socket, level, optname, optional_value.value());
+  } else {
+    return 0;
+  }
+}
+
+int TcpKeepaliveOptionImpl::setSocketOption(Socket& socket, int level,
+                                            Network::SocketOptionName optname, int value) {
+  if (!optname.has_value()) {
+    errno = ENOTSUP;
+    return -1;
+  }
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  return os_syscalls.setsockopt(socket.fd(), level, optname.value(), &value, sizeof(value));
```

Comment:
I'm going to change SocketOptionName to be absl::optional<std::pair<int, int>>, so it can contain the "level" as well as "optname." That should make it possible to clean up the socket option code a bit and maybe make a common Socket::Option implementation, which takes a SocketOptionName and value as constructor args, possible. That might then allow TCP Keepalive, and other simple options, to be implemented with as a factory that produces a vector of options.

SocketOptionImpl has special handling for checking if the socket it's being applied to is IP, and setting different options for IPV6 vs 4. I think that's going to have to remain slightly special (although it might be able to use the generic implementation under the hood.)

Reply:
Awesome. @bmetzdorf for visibility, so that you folks don't collide on this. I could merge #2793 as is if that works for you both.

Reply:
SGTM, I can address the open aesthetic items from #2793 in another PR.

Reply:
OK, will merge.

Reply:
https://github.com/envoyproxy/envoy/pull/2793 merged.